### ISSUE TITLE
Finetuning Script Feedback

### ISSUE THREAD

**Issue by MrMuhannadObeidat** [knottwill/sesame-finetune#2]
Hi Will, thanks for dropping this. It certainly much better than the lora approach as it gives more control for training and higher quality model. Few small observations:
1. The expected data format is pickle which does not support loading the data gradually. I understand it is fine from your perspective as you need to bin all similar examples together but it may present some scalability issues.
2. I have 1 million+ training samples and trying to put them into a single pickle file is a challenge. The code does not support running one big training loop across different files. I made the changes myself, but would be great if you can comment on this.
3. Lastly I am trying to train for a new language, Arabic. I have lots of examples of acceptable quality audio and transcripts, multi speaker. I also have a few thousand examples of high quality single speaker. I am doing two phases of training. First train with acceptable quality samples and then take the best_val model and train with single speaker high quality. The resulting model is not great and the loss only drops to 3.5 after all these training samples. On the contrary training Orpheus from Canopy produces a much better model. Same also for XTTS. Do you have any insights to why that might be. Anything in training parameters, hyper parameters or approach, that I should change?

**Comment by knottwill** [comment]
Hi Muhannad, 

Thanks for raising this - you make a good point about the pickle files not being scalable. Using h5py could be a good alternative as I know it supports partial loading. If you have made changes already it would be great if you could open a pull request and I will review / merge :)

For point 3, I suspect this is related to the tokenization. I am not sure how the Llama tokenizer deals with arabic but it might be very out-of-distribution for the Sesame CSM. It could also be that arabic is OOD for the Mimi tokenizer also. I have only finetuned into french and spanish so far, so have not encountered these problems.

**Comment by MrMuhannadObeidat** [comment]
Hello, appreciate your quick response on this.
I did not end up changing the code to another format as I saw you are doing binning of all samples to avoid padding. If the code was to change to load randomly from data set loader, then padding will be required and I did not have the details to proceed on that.

For point 3, I will validate the tokenzier from llama. I assumed Mimi will not require changes, not sure how to validate that.

**Comment by knottwill** [comment]
Ok no worries. If I find time then I will look into fixing this - keeping this issue open for now. 

As for the tokenization, you could finetune Mimi on your Arabic data but that would require significant effort. I recommend doing a robust sweep of finetuning params with my repo - that might be enough.

**Comment by knottwill** [comment]
I have now implemented pre-tokenization that iteratively saves the tokens in a `.hdf5` file, and partial loading from this dataset so we don't need to load everything into memory. 

Let me know if you have any issues.

**Comment by MrMuhannadObeidat** [comment]
Thanks a lot Will, this is great. I just cloned the repo and trying that.
I am also trying the sweep suggestion for training to find the right hyperparams but I need to reduce the data set size in that case as I need every trial to go through a resonable amount of data that does not go out of control.
